### ISSUE TITLE
Azure plugin: missing Azure Private zone prevents Coredns from starting up

### ISSUE THREAD

**Issue by iisworks** [coredns/coredns#4601]
**What happened**:
We are testing Coredns as a service on Ubuntu 18.04 in Azure, using the Azure plugin to load 50 Azure Private DNS zones. I noticed that when one of these zone resources gets removed from Azure while it is still being referenced in the Corefile, Coredns fails to start up completely. 

**Log**:
On an attempt to resolve anythinh in that zone I do see the expected error message while Coredns is still running:
_May 04 18:02:31 CoreDnsPrd000001 coredns[6725]: [ERROR] plugin/azure: Failed to update zones [ privatelink.afs.azure.net.]: errors updating zones: [failed to list resource records for privatelink.afs.azure.net from azure: privatedns.RecordSetsClient#List:_ 
But after stopping and restarting the Coredns service will not start at all, which is obviously pretty risky from an availability perspective.

**What you expected to happen**:
We expected Coredns Server to continue serving all other plugins/zones that could be loaded successfully, just generating an error for the failed plugin.

**How to reproduce it (as minimally and precisely as possible)**:
- Create 2 Private DNS Zone resources in Azure
- Configure the Azure plugin in the Corefile to use them
- Remove 1 Private Dns Zone resource from azure
- Coredns now fails to start

**Environment**:
- Coredns 1.8.3 on Ubuntu 18.04

**Comment by iisworks** [comment]
Just for my understanding: are bug reports usually confirmed or commented on by anyone from the team?

**Comment by chrisohaver** [comment]
Cc @miekg @yongtang @darshanime

**Comment by yongtang** [comment]
I am quite occupied next 2 weeks though I will take a look either after that, or before if I can find some extra time.

**Comment by yongtang** [comment]
@iisworks Can you provide a Corefile of the setup that encountered the issue?

**Comment by iisworks** [comment]
This is a (redacted) sample, although I don't think the issue is specific to the file (I tested with a much shorter list of Privatelink zones as well). 
[Corefile.txt](https://github.com/coredns/coredns/files/7098738/Corefile.txt)

We are using an Azure User "Managed Identity" on the VM scale set that allows the Coredns VMs to access the Private Dns Zone resources from the plugin.

Should be easy to repro as described before:
- Create 2 Azure Private Dns zones (Zone1 and Zone2)
- Add zones to the Corefile as valid Azure zones (using SPN + Key or Managed Identity)
- Run Coredns -> OK
- Remove the private dns zone Zone1 from Azure
- Coredns still runs, servering other Zone2
- Restart Coredns -> It completely failed to start! This means everything breaks when a VM instance in the scaleset reboots/reloads until the zone or config is corrected.
- Expected: Coredns should still start and serve all other plugins + Azure zones (Only log an error for failed Zone1)

**Comment by chrisohaver** [comment]
The azure plugin requires a successful update _before_ starting the update loop go routine, due to ...
https://github.com/coredns/coredns/blob/9b3b2679f5020f10cb8ef4c0bad33831333f09fa/plugin/azure/azure.go#L83-L85


Seems like the initial update could be moved to within the go routine, which would allow CoreDNS to start up ... e.g. 
```go
// Run updates the zone from azure.
func (h *Azure) Run(ctx context.Context) error {
	go func() {
		if err := h.updateZones(ctx); err != nil && ctx.Err() == nil {
			log.Errorf("Failed to update zones %v: %v", h.zoneNames, err)
		}
		for {
			select {
			case <-ctx.Done():
				log.Debugf("Breaking out of Azure update loop for %v: %v", h.zoneNames, ctx.Err())
				return
			case <-time.After(1 * time.Minute):
				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil {
					log.Errorf("Failed to update zones %v: %v", h.zoneNames, err)
				}
			}
		}
	}()
	return nil
}
```

**Comment by iisworks** [comment]
So that would mean it only skips the missing Azure zone, but will successfully start Coredns and load all other valid Azure zones and other plugins (Forward)? That sounds a lot more resilient! :)

**Comment by chrisohaver** [comment]
> So that would mean it only skips the missing Azure zone, but will successfully start Coredns and load all other valid Azure zones and other plugins (Forward)? That sounds a lot more resilient! :)

Yes.  Some other plugins that connect to external resources to serve a zone do it this way.

**Comment by chrisohaver** [comment]
Actually, skimming the azure plugin code over (I'm not familiar with it) other changes might need to be made e.g. in `setup`, and `New()` - which could require some non-trivial refactoring.

**Comment by iisworks** [comment]
Thanks for looking into it! We obviosuly would really appreciate having more resiliance, especially with the increased risk of all Coredns instances failing to start at all. We use many (50+) separate Azure Private Link zones and other Azure internal dns zones that theoretically could inadvertently be removed or encounter a transient Azure issue. 

From our perspective it still looks more like a bug than just an enhancement to be honest. But either way, how is decided if this gets picked up in the (near) future?

**Comment by chrisohaver** [comment]
> But either way, how is decided if this gets picked up in the (near) future?

Anyone is welcome to pick it up and open a PR at any time.  Once it's approved/merged, it will be included in the following release.

**Comment by iisworks** [comment]
I see, and I would if I could. :)

**Comment by houyuwushang** [comment]
@yongtang I reproduced this on current master with the real Azure SDK against a local HTTP test server, for both private and public zones. A missing zone fails both the constructor's remote validation and the initial synchronization. A later failed refresh also replaces a previously valid snapshot with an empty zone.

I'll prepare a focused fix that starts cloud synchronization in the background, isolates errors per zone, and publishes only complete successful snapshots. Previously loaded data will remain available during failed refreshes; never-loaded zones will return SERVFAIL unless explicit fallthrough is configured. The tests will cover missing-zone startup, healthy sibling zones, pagination errors, recovery and shutdown cancellation. I'll keep the SDK/authentication migration in #6864 separate.
